app/code/Magento/Catalog/Controller/Adminhtml/Product/Save.php

```diff
@@ -117,6 +117,7 @@ public function execute()
                 $productId = $product->getEntityId();
                 $productAttributeSetId = $product->getAttributeSetId();
                 $productTypeId = $product->getTypeId();
+                $this->getRequest()->setParam('type', $productTypeId);
```

Comment:
I cannot add more on this :) Maybe some common principle, like CQRS, exists and it says that we should not modify what we received in request.

In general, the fewer mutations - the better. For example, in https://github.com/magento/magento2/blob/develop/app/code/Magento/Catalog/Model/Product/TypeTransitionManager.php#L48 product type changing responsibility is clearly encapsulated. Here we just modify a request in arbitrary place.